src/duckling/game/collision/collision-types.service.ts

```diff
@@ -0,0 +1,164 @@
+import {JsonLoaderService} from '../../util/json-loader.service';
+import {ProjectService} from '../../project/project.service';
+import {SnackBarService} from '../../project/snackbar.service';
+
+import {getCollision} from './collision-attribute';
+
+export type CollisionTypeGetter = (entity : Entity) => string;
+
+interface CollisionTypeMetaData {
+    collisionTypes: string
+}
+
+/**
+ * This service can be used to get the collision types for an entity or all unique collision types in an entity system
+ */
+@Injectable()
+export class CollisionTypesService {
+    collisionTypes = new BehaviorSubject<string[]>([]);
+    
+    constructor(private _jsonLoader : JsonLoaderService,
+                private _store : StoreService,
+                private _project : ProjectService,
+                private _snackbar : SnackBarService) {
+        this._store.state.subscribe(state => {
+            if (state.collision.collisionTypes) {
+                this.collisionTypes.next(state.collision.collisionTypes);
+            }
+        });
+    }
+    
+    /**
+     * Get the collision type for a specific entity
+     */
+    getCollisionTypeForEntity(entity : Entity) : string {
+        return this._collisionType(entity);
+    }
+
+    /**
+     * Get all the unique collision types for an entity system
+     */
+    getUniqueCollisionTypesInEntitySystem(entitySystem : EntitySystem) : string[] {
+        let collisionTypes = new Set<string>();
+        entitySystem.forEach((entity : Entity, key : EntityKey)  => {
+            let collisionType = this.getCollisionTypeForEntity(entity);
+            if (collisionType) {
+                collisionTypes.add(collisionType);
+            }
+        });
+        return Array.from(collisionTypes.values());
+    }
+
+    /**
+     * Add a new collision type to the game's collision types
+     */
+    addCollisionType(collisionType : string) {
+        this._store.dispatch(_collisionTypesAction(Array.from(this.collisionTypes.getValue().concat([collisionType]))));
+    }
+
+    /**
+     * Registers the game collision types from the project/collision-types.json file during
+     * the map's after load lifecycle.
+     */
+    async registerGameCollisionTypes(map : RawMapFile) : Promise<RawMapFile> {
+        let json = await this._jsonLoader.getJsonFromPath(this._project.getMetaDataPath("collision-types"));
+        let collisionTypes = new Set<string>(["none"]);
+        if (json) {
+            let collisionTypeMetaData : CollisionTypeMetaData = JSON.parse(json);
+            for (let collisionType of collisionTypeMetaData.collisionTypes) {
+                if (collisionType !== "none") {
+                    collisionTypes.add(collisionType);
+                }
+            }
+        }
+        
+        collisionTypes = this._findUnknownCollisionTypes(map, collisionTypes);
+        this._store.dispatch(_collisionTypesAction(Array.from(collisionTypes)));
+        this._saveCollisionTypesMetaData();
```

Comment:
The logic here is to save the collision types that were unknown (found in the map file but not the metadata file) as soon as we register them. Should we do this or save when we hit save?